selfdrive/locationd/models/live_kf.py

```diff
@@ -193,96 +207,37 @@ def generate_code(generated_dir):
-    self.filter.init_state(state, P, filter_time)
-
-  def predict_and_observe(self, t, kind, meas, R=None):
-    if len(meas) > 0:
-      meas = np.atleast_2d(meas)
-    if kind == ObservationKind.CAMERA_ODO_TRANSLATION:
-      r = self.predict_and_update_odo_trans(meas, t, kind)
-    elif kind == ObservationKind.CAMERA_ODO_ROTATION:
-      r = self.predict_and_update_odo_rot(meas, t, kind)
-    elif kind == ObservationKind.ODOMETRIC_SPEED:
-      r = self.predict_and_update_odo_speed(meas, t, kind)
-    else:
-      if R is None:
-        R = self.get_R(kind, len(meas))
-      elif len(R.shape) == 2:
-        R = R[None]
-      r = self.filter.predict_and_update_batch(t, kind, meas, R)
-
-    self.filter.normalize_state(States.ECEF_ORIENTATION.start, States.ECEF_ORIENTATION.stop)
-    return r
-
-  def get_R(self, kind, n):
-    obs_noise = self.obs_noise[kind]
-    dim = obs_noise.shape[0]
-    R = np.zeros((n, dim, dim))
-    for i in range(n):
-      R[i, :, :] = obs_noise
-    return R
-
-  def predict_and_update_odo_speed(self, speed, t, kind):
-    z = np.array(speed)
-    R = np.zeros((len(speed), 1, 1))
-    for i, _ in enumerate(z):
-      R[i, :, :] = np.diag([0.2**2])
-    return self.filter.predict_and_update_batch(t, kind, z, R)
-
-  def predict_and_update_odo_trans(self, trans, t, kind):
-    z = trans[:, :3]
-    R = np.zeros((len(trans), 3, 3))
-    for i, _ in enumerate(z):
-        R[i, :, :] = np.diag(trans[i, 3:]**2)
-    return self.filter.predict_and_update_batch(t, kind, z, R)
-
-  def predict_and_update_odo_rot(self, rot, t, kind):
-    z = rot[:, :3]
-    R = np.zeros((len(rot), 3, 3))
-    for i, _ in enumerate(z):
-        R[i, :, :] = np.diag(rot[i, 3:]**2)
-    return self.filter.predict_and_update_batch(t, kind, z, R)
+    # this returns a sympy routine for the jacobian of the observation function of the local vel
+    in_vec = sp.MatrixSymbol('in_vec', 6, 1)  # roll, pitch, yaw, vx, vy, vz
+    h = euler_rotate(in_vec[0], in_vec[1], in_vec[2]).T*(sp.Matrix([in_vec[3], in_vec[4], in_vec[5]]))
+    extra_routines = [('H', h.jacobian(in_vec), [in_vec])]
+
+    gen_code(generated_dir, name, f_sym, dt, state_sym, obs_eqs, dim_state, dim_state_err, eskf_params, extra_routines=extra_routines)
+
+    # write constants to extra header file for use in cpp
+    live_kf_header = "#pragma once\n\n"
+    live_kf_header += "#include <unordered_map>\n"
+    live_kf_header += "#include <eigen3/Eigen/Dense>\n\n"
+    for state, slc in inspect.getmembers(States, lambda x: type(x) == slice):
+      assert(slc.step is None)  # unsupported
+      live_kf_header += f'#define STATE_{state}_START {slc.start}\n'
+      live_kf_header += f'#define STATE_{state}_END {slc.stop}\n'
+      live_kf_header += f'#define STATE_{state}_LEN {slc.stop - slc.start}\n'
+    live_kf_header += "\n"
+
+    for kind, val in inspect.getmembers(ObservationKind, lambda x: type(x) == int):
+      live_kf_header += f'#define KIND_{kind} {val}\n'
```

Comment:
can we use enums here?

Reply:
Changed the name to OBSERVATION, enums would have to be converted to ints when using the rednose library, so I kept it with defines